design-2.md

```diff
@@ -0,0 +1,119 @@
+# Design-2
+
+A simple but bad program for the empty room would be as follows:
+
+    action go right
+    can move right {
+        move right
+        go right
+    }
+    action go up
+    can move up {
+        move up
+        go up
+    }
+
+    action go left
+    can move left {
+        move left
+        go left
+    }
+    action go right 2
+    can move right {
+        move right
+        go right 2
+    }
+    move down
+    go left
+
+Here is a program using abstractions:
+
+    action move all the way (dir) {
+        action continue
+        can move dir {
+            move dir
+            continue
+        }
+    }
+    control loop (body) {
+        action continue
+        body
+        continue
+    }
+    
+    move all the way right
+    move all the way up
+    loop {
+        move all the way left
+        move all the way right
+        move down
+    }
+    
+The basic commands are "can", "can't", "move", "action", and "control". Action defines functions.
+In particular, action with just an identifier defines a function which can be used as a "goto".
+If an action does take in an argument, it is a direction.
+Control defines a control flow structure, which takes in a body.
+There's probably a nicer way to combine action and control, but it might require type annotation.
+
+States are simply implicitly created in between lines basic lines (like "move right"),
+so this is fairly state-wasteful in the code it generates.
+
+A final trick is that you can also use where the picobot is facing. For example, for a maze:
+
+    move all the way to your right
+    loop {
+        can move to your right {
+            face to your right
+        }
+        move forward
+    }
+
+Since the direction picobot is facing has to be included in the state info, using this can
+cause and additional x4 blowup in produced number of states.
+
+## Who is the target for this design, e.g., are you assuming any knowledge on the part of the language users?
+
+The target is essentially programmers (or learners), because the thought processes that go into
+using the language, picobot in general, and especially defining new commands in the language are targeted
+at programmers. However, it is designed so that it might be readable by non-programmers.
+
+## Why did you choose this design, i.e., why did you think it would be a good idea for users to express the maze-searching computation using this syntax?
+
+We wanted to capture the essence of what someone is actually thinking when they try to write a
+picobot program. It has a lot to do with general, higher-level actions,
```

Comment:
I'd argue that a programmer isn't thinking about goto's when writing a picobot solution.